### DISCUSSION TITLE
Improve Go release process

### DISCUSSION THREAD

**Discussion by andig** [Homebrew/discussions#2618] [Tap maintenance and Homebrew development]
Follow-up to https://github.com/Homebrew/homebrew-core/pull/90350

**Comment by andig** [comment]
/cc @SMillerDev @stefanb

> Yes, it is chicken and egg problem in a way, but current approach is blocking gradual evolution progress.

I fully agree. The problem is that we're penalizing everyone plus security. I also don't quite grasp the need for doing so. The current binaries for compiled applications continue to get distributed.

We should try to move the fixing process for applications to their releases, not to the platform (i.e. Go) release.

**Comment by iMichka** [comment]
Wondering if we could take the same approach as for Python. This means versioning the go formulate completely, to be able to always ship the latest release (we do ship Python@3.10 right now, but the migration of dependants is yet not finished).
This has the advantage to make all the go users happy as now we can instantly ship the newest, shiny go version.

The only drawback I see is that all our packages will stay on old version of go and we will always lag behind, thus not benefiting from the latest security updates. I also do not know if the full go stack needs to be in sync, or if you can work with mismatched go library versions.

**Reply by carlocab** [reply]
Migrating Python formulae is a huge pain though. I'd really like to not have to do the same for all the Go formulae.

**Comment by kidoman** [comment]
There are two driving forces at odds here.

* The Go community loves the latest version because we have gotten to a point of confidence of the new version not breaking anything (almost always.) So we want to get the latest version to enjoy the performance/security benefits.

* The brew maintainers do not want a lot of tech debt to set in. If the process of upgrading versions is not iterative, then they would have to make a bug push to eventually upgrade the Go compiler version.

I love my new Go releases. Every time there is a new Go release, one of my first reactions is `brew upgrade go`. I fondly remember when the Go releases would come up on `brew` almost immediately - sometimes the day the Go team would release the patch.

I am sure everyone would agree that the shiny new versions of Go being available on brew is not a bad thing per-say. After all, `brew` as the de facto package manager for so many people, would be serving its purpose super well if that deltaT was ~0.

With that said, how can we make the `go` formula track the `go` releases more closely - or less painfully?

**Comment by carlocab** [comment]
Personally, I wouldn't mind skipping build testing Go dependents on minor Go releases. This should help us get security/bug fixes out more quickly.

Major releases should still do (at least some) build testing. I'm open to this being a fraction of Go-dependent formulae, rather than just all of them.

**Comment by andig** [comment]
> The brew maintainers do not want a lot of tech debt to set in. If the process of upgrading versions is not iterative, then they would have to make a bug push to eventually upgrade the Go compiler version.

I still don't fully understand exactly who has the technical debt here. I app A doesn't work with new Go version it will still work with old Go version. If brew is able to deliver that binary compiled with the old version that's find. The next app release will be painful, but only for that app instead of the entire go community.

**Comment by kidoman** [comment]
Concur. Do we need someone to unblock this from being stalled in process?

**Comment by iMichka** [comment]
> I still don't fully understand exactly who has the technical debt here. I app A doesn't work with new Go version it will still work with old Go version. If brew is able to deliver that binary compiled with the old version that's find. The next app release will be painful, but only for that app instead of the entire go community.

It's a little bit more complex:
- You might have what I call "dependency diamonds": Package A depends on `go` 1.17.5, package B on `go` 1.17.4. Package C depends on both package A and B: we are stuck now. Not sure if this happens a lot in the `go` ecosystem. But let's ignore this and do as you say, let's just leave A, B, and C on legacy `go` 1.17.4.
- Let's also just ignore all the reverse dependencies now, so basically let's leave all the packages from the following list on `go` 1.17.4: `brew uses go --include-test --include-build --recursive`. That's roughly 700 packages.
- Fast forward 2 years later. We are now shipping `go` 1.37.0. From the 700 packages, 250 are broken, because nobody cared.

Now comes the thing; there is a critical `openssl` update to be pushed. 250 packages are broken because nobody cared. If we have fixed them from time to time, we would have been able to push the `openssl` update. So now we are stuck.

It's not just about `go`. We could use the same "we don't care" strategy for Python, Rust, Lua, and any other critical package. You see that the list of 250 broken packages will quickly grow.

We have seen this scenario happen. Multiple times. It's called technical debt.
We need to update things as they break, not let things pile up. I know that most companies in the tech industry do not care about technical debt, but we do. And we care about all the packages we ship. We are a package manager: all packages need to work.

Also: we made a few users happy because we shipped their latest `go` release, but we have now the users from 250 packages that are unhappy because their packages does not work anymore. That's potentially more people than the people that wanted a `go` update.

**Reply by andig** [reply]
> > I still don't fully understand exactly who has the technical debt here. I app A doesn't work with new Go version it will still work with old Go version. If brew is able to deliver that binary compiled with the old version that's find. The next app release will be painful, but only for that app instead of the entire go community.
> 
> It's a little bit more complex: ...

> Also: we made a few users happy because we shipped their latest `go` release, but we have now the users from 250 packages that are unhappy because their packages does not work anymore. That's potentially more people than the people that wanted a `go` update.

But why is that? Those 250 packages can and are still distributed as working binaries. There is no reason to rebuild them on the latest Go (or any other toolchain)?

> Now comes the thing; there is a critical `openssl` update to be pushed. 250 packages are broken because nobody cared. If we have fixed them from time to time, we would have been able to push the `openssl` update. So now we are stuck.
> 
> It's not just about `go`. We could use the same "we don't care" strategy for Python, Rust, Lua, and any other critical package. You see that the list of 250 broken packages will quickly grow.
> 
> We have seen this scenario happen. Multiple times. It's called technical debt. 

Absolutely. I think this boils to to the question of what is the scope and goal of Brew. If Brew is the missing package manager, then it would in my eyes to be totally acceptable and expected to distribute packages of _working_, i.e. maintained software. If, on the other hand, Brew want to provide the broadest possible software distribution then it must also take over the burden of fixing the zillion abandon packages.

As for me, Brew as package manager for maintained software is good enough 👍🏻

**Reply by iMichka** [reply]
> But why is that? Those 250 packages can and are still distributed as working binaries. There is no reason to rebuild them on the latest Go (or any other toolchain)?

Yes, binaries will still work (I hope, I don't know if `go` binaries are self-contained/do not link against other stuff?). I see that most of `go` packages do not depend that much on other things. `go` is a little bit more special here. Other core packages do need regular revision bumps and rebuilds of their recursive dependants, whereas `go` seems to not have that issue.

But we consider more things as "breakage" than just "the binary does not work". For example an upstream source checksum change is a major security issue and flags the package as "broken". Upstream urls disappearance is also another concern.
We also have tests in place that check if anything gets broken during an update. We also want to keep these test all green.
Ignoring these are just leaving red CI forever, which sort of means we don't need CI at all.

I admit that `go` is maybe a special case, compared to some other packages, as reverse dependants will not get broken that often. But we also would like to keep the same rules and workflows for all packages.

Finally, we also use the "broken" reverse dependents test to force users to submit fixes. You want a new `go` version, please help fix things. From our point of view it's a good strategy because it has helped fixing a lot of things which would have stayed broken if not. But on the other side I can understand the frustration, especially when the list of broken things grows, and you are the only one submitting patches. Sometimes it feels like emptying the ocean with a spoon. 

> As for me, Brew as package manager for maintained software is good enough

Yes, we want to ship only maintained software. We have lately introduced `deprecate!`  and `disable!` stanzas to slowly phase out broken packages. I plan to use these a little bit more often to stop carrying broken stuff around. We also made some adjustments on what get's tested to transform a few errors in warnings.

So right now the situation is probably not going to improve. I will try to bring up the issue at the next maintainer meeting to see if we can find some more improvements.

**Reply by gonzaloserrano** [reply]
Hi 👋 

First, I agree and support brew's goal to maintain healthy formulae.

> I don't know if go binaries are self-contained/do not link against other stuff

Yep they are/they do not link against anything.

See https://go.dev/doc/faq#Why_is_my_trivial_program_such_a_large_binary

> You want a new go version, please help fix things.

The thing here is that even the main language of these projects is Go, 99% of the time what's broken is not go-related: e.g some formula issue, infra stuff or frontend stuff. This is because of Go's efforts to make newer versions not break old ones, also known as "Go compatibility promise" - https://go.dev/doc/go1compat

Let's take a look at the fixes for formulas broken in the Go upgrade PR - https://github.com/Homebrew/homebrew-core/pull/90350#issuecomment-991958277:
- vgrep: fixed formula test asserts
- vault 1.9.1: fixed formula
- traefik: some yarn build issues (js/node)
- skaffold: fixed formula checksum
- nuclei: same thing
- mmctl: downgraded because latest version is missing
- charm-tools: not sure why this is triggered for the go formula since it's a python package?

Other failures:
- chronograf: `error /private/tmp/chronograf-20211211-95302-1qnqr0s/chronograf-1.9.1/node_modules/deasync: Command failed.` (node/js)
- akamai: some formulae test asserts failing `Minitest::Assertion: Expected /Purge/ to match "\n    `
- awsweeper: some terraform stuff failing `Minitest::Assertion: Expected /Error:\ failed\ to\ configure\ provider\ \(name=aws/ to match "\nfailed to install provider (aws): 2 problems:\n\n- registry.terraform.io: This version of Terraform has an outdated GPG key and is unable to verify new provider releases. Please upgrade Terraform to at least 0.12.31 to receive new provider updates. `
...

> But we also would like to keep the same rules and workflows for all packages.

None of these are broken because a new Go formula version was upgraded, but because those projects were rebuilt. Probably they would have broken with the previous (or any) Go version too. 

Maybe there should be a (semi)automated way to detect those? E.g rebuild every X time with the same versions and detect these failures earlier. Let's say a project works with Go 1.17.4, and later it fails to build with the same version 1.17.4 (because it's triggered with some Github Actions schedule event or whatever). It could be marked with some stanza that so if a new Go version 1.17.5 is published, it would not unblock the new release.

**Reply by XuehaiPan** [reply]
>> You want a new go version, please help fix things.
>
> The thing here is that even the main language of these projects is Go, 99% of the time what's broken is not go-related: e.g some formula issue, infra stuff or frontend stuff. This is because of Go's efforts to make newer versions not break old ones, also known as "Go compatibility promise" - https://go.dev/doc/go1compat
>
> None of these are broken because a new Go formula version was upgraded, but because those projects were rebuilt. Probably they would have broken with the previous (or any) Go version too.

Yes, sometimes the CI test fails because the downstream formula is broken anyway (e.g. invalid resource URL).

Can we automatically rerun CI tests for the corresponding formula with the previous version of `go` in a `go` version migration pull request? Same strategies for some core formulae (e.g. `python@3.x`).

**Comment by iMichka** [comment]
> Concur. Do we need someone to unblock this from being stalled in process?

Not one single person can do this. We need more maintainers. We need more people that fix stuff. Even the easy things. This gives us more time to fix the hard parts.

Just have a look at homebrew-core and fix anything that has red CI.

**Comment by Moulick** [comment]
> > Also: we made a few users happy because we shipped their latest go release, but we have now the users from 250 packages that are unhappy because their packages does not work anymore. That's potentially more people than the people that wanted a go update.

> But why is that? Those 250 packages can and are still distributed as working binaries. There is no reason to rebuild them on the latest Go (or any other toolchain)?

Thing is, that might hold true all the time. What happens when there is need to rebuild some binary for any reason? Maybe they get lost, corrupted, or maybe Brew needs to shift the binary storage (happened recently) and some reason needs to recompile every go package. That would not work.
I'd rather have a `go` version in brew that can build all the `go` dependent packages in brew and be somewhat recent, rather than the absolute latest version that can only build some of them. 
There are many ways to install the latest version of `go` till brew updates it, like `go install golang.org/dl/go1.17.5@latest`

**Reply by andig** [reply]
> I'd rather have a go version in brew that can build all the go dependent packages in brew and be somewhat recent, rather than the absolute latest version that can only build some of them.

That's exactly the anti-point of the discussion above. Go basically guarantees that. Almost none of the failures are related to Go but to anything else, either formula- or package-related.

**Reply by Moulick** [reply]
In a way yeah. The absolute best would be every package is rebuilt nightly to catch any such issues. But that cannot happen due to crazy amount of resources that'll be used/wasted. 
Running a full rebuild for all packages that happens on a new `go` version bump, even if the issue is not due to the build failing due to `go` specifically, atleast gives a major incentive to get most of such issues fixed.

Afaik when 1.16 was released, I think it took more than about a week to get it through. But the number of fixes that generated was huge. Lotta things got fixed, even if the release was delayed on brew for about a week https://github.com/Homebrew/homebrew-core/pull/71289

**Reply by SMillerDev** [reply]
> Go basically guarantees that. Almost none of the failures are related to Go but to anything else, either formula- or package-related.

But that's kind of the problem here. Despite the language's best intentions Go programs turn out to be not so reproducible as designed. You simply can't take the language promise and apply it to a package management context.

**Reply by bmhatfield** [reply]
@SMillerDev I'm a bit confused by that statement. Following from afar, plus other comments here, I've formed the understanding that the issues resolving builds _almost never have to do with changes to Go_, its general entropy in the ecosystem, with tons of broken formulas, broken builds, etc. This is what folks are pushing back against: [Go changes aren't breaking these homebrew builds](https://github.com/Homebrew/homebrew-core/pull/94895#issuecomment-1049066162). They were already broken for other reasons.

**Reply by SMillerDev** [reply]
>  I've formed the understanding that the issues resolving builds almost never have to do with changes to Go

That's what people keep telling me, but it's hard to verify since they do sometimes randomly stop working, which is a problem no matter what.

My view on the matter is this: Go as a language promises things that aren't applicable in a package manager context and instead of helping to fix the broken software and resolve the issues the Go community prefers to yell at Homebrew maintainers making me very demotivated and ensuring that Go will always be last on my list of priorities.

**Reply by stefanb** [reply]
> That's what people keep telling me, but it's hard to verify since they do sometimes randomly stop working, which is a problem no matter what.

It is very easy to verify this: try re-building a dummy PR with same Go version as currently on homebrew (1.17.6 at the moment) it will likely fail with some random formula, possibly the same as in PR for "problematic" 1.17.7 or 1.18rc1.

**Reply by iMichka** [reply]
We did have broken `go` packages in the past when bumping minor versions. That's why all these checks are in place. And breakages were related to `go` changes, not unrelated stuff.

It just happens that unrelated breakages also happen. These need also to be fixed. But no one seems to care, except maintainers. And that's a big issue in my opinion.

We have a green CI policy. For `go`, for `Python`, for `rust`, and all the other 5500 packages we ship. Why would we make an exception for `go`?

There is a small group of dedicated people that try to keep things working and actually fix packages, while 100 other are complaining. Look at this discussion. It's endless. So much energy has been put into answering. And not that much energy has been put into fixing things.

**Reply by andig** [reply]
> It is very easy to verify this: try re-building a dummy PR with same Go version as currently on homebrew (1.17.6 at the moment) it will likely fail with some random formula, possibly the same as in PR for "problematic" 1.17.7 or 1.18rc1.

> There is a small group of dedicated people that try to keep things working and actually fix packages, while 100 other are complaining. Look at this discussion. It's endless. So much energy has been put into answering. And not that much energy has been put into fixing things.

Let is please not go that path down again. We've been there, everyone lost.

Parts of the go community care about the compiler only, Homebrew cares about all packages and wants them fixed. For my personal taste Homebrew carries too much outdated software and tries to keep things alive that are already decaying, but that's opinion. Love it or leave it.

My suggestion above was making the fixing easier as tasks are often similar (releases, missing upstream attention etc). Doing this in Ruby is beyond me.

**Reply by SMillerDev** [reply]
> Parts of the go community care about the compiler only, Homebrew cares about all packages and wants them fixed.

I think the best option for both sides is someone in the Go community hosting an unnoficial tap with bleeding edge Go versions. Devs get to use whatever version they want and Homebrew gets to do package management beyond just Go.

> For my personal taste Homebrew carries too much outdated software and tries to keep things alive that are already decaying, but that's opinion. Love it or leave it.

We are very actively deprecating software all the time, if software is in Homebrew it's because we can't find any indication of it being deprecated/abandoned.

**Comment by danilobuerger** [comment]
I am guessing that things will get a little bit heated once go 1.18 is released as everybody wants to get their hands on generics.

**Reply by Moulick** [reply]
`go install golang.org/dl/go1.18.0@latest` ? Although I consider brew to be my main source of packages, it's not my only source. 
It's almost as easy to get the version from golang's official website for a few days, till brew updates.

**Comment by iMichka** [comment]
What could also be done is to open a pre-build pull request a few days before the `go` release. Do they have RC candidates?

For Python for example there is often someone opening a RC pull request, to run a first build and fix any issue before the real pull request is pushed. This has also proven to be quite efficient.

Also, I do not think Homebrew are the slowest ones to ship `go` updates. 1 week delay is already quite a tight schedule (unless there is major security fix to push, then we can be way faster).

**Reply by kidoman** [reply]
This is a legit idea and I think has a lot of merit. They do have multiple RC candidates, but that is usually for the major releases (the next one will be 1.18.) The patches (like go 1.17.5) come out without official RCs, but I am not too sure.

If we open the PR on the last couple of RCs, we should be in pretty good shape for the final release.

I would love to help, but finding the whole thing a bit daunting (been a user of `brew` for years, and I want to give back.) Maybe some of us can use the 1.18 release as a learning experience.

**Reply by iMichka** [reply]
At least opening the pull request, letting CI run, and make a list of errors we can check is already giving us a small heads up. Some errors are easily fixable by newcomers and we can guide you/them. Some others need a little bit more experience, but we can handle that.

**Reply by kidoman** [reply]
Thanks. That's encouraging. I will keep a look out for 1.18 RCs.

**Reply by kidoman** [reply]
Thinking out loud: Should we get the process start (go 1.18 is expect to be out in GA by Feb 2022) by compiling golang's master branch to start discovering issues? 1.18 is expect to big language change, and I expect us to discover rough edges through this process.

**Reply by iMichka** [reply]
Maybe it's a little bit early now. A few weeks before would be good. Christmas period is coming so things might slow down a little bit for a few maintainers (and for upstream packages too :))

**Reply by kidoman** [reply]
Makes sense. Will keep the PR ready when its time.

**Reply by Moulick** [reply]
I think I saw @andig open a RC build PR for 1.17 as well to do exactly the same, find out the issues before 1.17 lands

**Comment by iMichka** [comment]
Looking at https://go.dev/doc/go1compat, upstream seem's to say: we will try to not break anything on best effort, but don't promise anything either. So recompiling everything each time is the best thing to do. I think we have different ways of interpreting that document.

Even if right now all the issues we did pick up were unrelated, I bet we will pick up a legit issue one day. And we probably already did in the past (I do not have any example to back this up, might need to dig into the history).

**Reply by kidoman** [reply]
Would it be a bad idea to use the Go 1.18 release as a learning experience and make a decision based on that? I can definitely help curate the data required for the decision (how many things broken with a big Go release with massive language changes).

**Reply by SMillerDev** [reply]
What do you suggest would be decided based on that release?

**Comment by andig** [comment]
> Would it be a bad idea to use the Go 1.18 release as a learning experience and make a decision based on that?

I feel we're starting to mix up two different things here and need to differentiate.

In https://github.com/Homebrew/discussions/discussions/2618#discussioncomment-1799775 it was pointed out that Go maintains compatibility at source code level and that minor releases typically don't break anything. The breakages observed are almost always related to formula or upstream problems. 

> Thing is, that might hold true all the time. What happens when there is need to rebuild some binary for any reason? Maybe they get lost, corrupted, or maybe Brew needs to shift the binary storage (happened recently) and some reason needs to recompile every go package. That would not work.

Those are introduced by compiling and can be solved when bumping upstream. They simply don't depend on Go. See https://github.com/Homebrew/discussions/discussions/2618#discussioncomment-1792803 for more argument.

I understand @iMichka will bring up this topic at the maintainer's meeting:

>  will try to bring up the issue at the next maintainer meeting to see if we can find some more improvements.

The other thing that you're referring to is major version (which in Go is .16, .17, .18 etc). Those builds have a higher chance of breaking due to changes in the `go` command, module handling (`-d`), deprecation warnings during linting etc. Those need be fixed, but that's not what I wanted to discuss specifically in this issue.

**Comment by stefanb** [comment]
Go 1.18beta1 was released yesterday and I created a PR https://github.com/Homebrew/homebrew-core/pull/91369 to test it early, as discussed previously.

**Reply by stefanb** [reply]
Currently [failing `tap_syntax`](https://github.com/Homebrew/homebrew-core/runs/4531570511?check_suite_focus=true) with:
> Stable version URLs should not contain beta

**Reply by SMillerDev** [reply]
https://github.com/Homebrew/discussions/discussions/2618#discussioncomment-1801672

Beta really seems too early. RC is a better target for Homebrew (but you can definitely ask upstream projects to add the beta to their CI as a canary test)

**Comment by rmetzler** [comment]
go 1.18 was released on 2022-03-15. https://go.dev/blog/go1.18

Despite there being multiple PRs for it, it still can't be installed through homebrew as of today on 2022-03-17. There are multiple other versions, but not the latest one. 

```
$ brew install go@1.18
Running `brew update --preinstall`...
Warning: No available formula with the name "go@1.18". Did you mean go@1.17, go@1.16, go@1.15, go@1.14 or go@1.13?
==> Searching for similarly named formulae...
These similarly named formulae were found:
go@1.17                                         go@1.16                                         go@1.15                                         go@1.14                                         go@1.13
To install one of them, run (for example):
  brew install go@1.17
```

```
$ brew upgrade go
# ...
Warning: go 1.17.8 already installed
```

Why can't we just have formulas with versions and one without? So that users and projects can decide what they want to install or add as dependency? We could have had the go1.18beta1 version for a few months already (when the first PRs were created), to test if projects compile with 1.18.

**Reply by SMillerDev** [reply]
> We could have had the go1.18beta1 version for a few months already (when the first PRs were created), to test if projects compile with 1.18.

We had a test PR open, it revealed issues that nobody fixed. Please actually read the responses here because all your ideas are answered.
